Add support for Spotify web player

JamesVStone opened the pull request:
I have added methods to communicate  track information from the chrome extension [here](https://github.com/SwagLyrics/SwagLyrics-Chrome-Extension). 

This relies on this pull request being merged into the chrome extension. [here](https://github.com/SwagLyrics/SwagLyrics-Chrome-Extension/pull/2).

pep8speaks commented:
Hello @JamesVStone! Thanks for opening this PR. 





* In the file [`tests/test_spotify.py`](https://github.com/SwagLyrics/SwSpotify/blob/e8e344d2077787d23d1b9d82e7adb7d5dfb2cefa/tests/test_spotify.py):

> [Line 201:9](https://github.com/SwagLyrics/SwSpotify/blob/e8e344d2077787d23d1b9d82e7adb7d5dfb2cefa/tests/test_spotify.py#L201): [F841](https://duckduckgo.com/?q=pep8%20F841) local variable 'x' is assigned to but never used
> [Line 209:9](https://github.com/SwagLyrics/SwSpotify/blob/e8e344d2077787d23d1b9d82e7adb7d5dfb2cefa/tests/test_spotify.py#L209): [F841](https://duckduckgo.com/?q=pep8%20F841) local variable 'x' is assigned to but never used

Do see the [Hitchhiker's guide to code style](https://goo.gl/hqbW4r)

aadibajpai commented:
You've bundled the extension here as well?

aadibajpai commented:
Additionally, can you explain exactly how everything works so it is easier and faster to review?

JamesVStone commented:
sure! Right so my solution consists of two pull requests (this one and one to the chrome extension which was referenced in the description).

@aadibajpai 
# How it works

I have added a file called `web_player.py` which spawns a web server which will last for 2 seconds or until it receives a post request from the chrome plugin.

This data is than passed to `get_info_chrome` method which allows the track and artist collected from Spotify Web Player to be included in the SwSpotify library. 

My modification to the chrome plugin makes it send the data more frequently so we can capture it more efficiently in this program. 

# Tests

I still need to improve coverage of the new code which I have added as the bot is telling me.

As for functionality it should work on all supported platforms (Windows, Mac and Linux). 

I have tested it progressively with SwagLyrics which works with the new version of my program. It successfully gets the track information and collects the lyrics!

JamesVStone commented:
@aadibajpai All finished. Fully threaded implementation which has timeout and memory safe fallback to complete. Apologies for my early submission. There are no dangling threads in the latest commit.

JamesVStone commented:
To clarify this PR requires on [these changes to the chrome extension](https://github.com/SwagLyrics/SwagLyrics-Chrome-Extension/pull/2).

aadibajpai commented on SwSpotify/spotify.py:
```diff
@@ -1,5 +1,7 @@
 import sys
+import os
 from SwSpotify import SpotifyClosed, SpotifyPaused
+from . import spotify_web
```
from SwSpotify import spotify_web is preferable so just tack it onto the previous line

aadibajpai commented on SwSpotify/spotify.py:
```diff
@@ -112,13 +114,31 @@ def get_info_mac():
+    """
+    Uses chrome extension to get the data
+    """
+    sys.stderr = open(os.devnull, "w")
```
You redirect stderr to devnull but I don't think you restore it?

JamesVStone replied:
would this restore it:
```python
sys.stderr = sys.__stderr__
```

JamesVStone replied:
I believe this is in order so I will mark as resolved for tm. Unresolv is it is still a problem.

aadibajpai replied:
I think it should too, btw why do you want to suppress errors here?

JamesVStone replied:
Because the base library which I am using `http.server` prints all requests to the server to stderror. I could not find an easy way to suppress this effect by overriding the class. If you have a better idea mention it as this is properly not the best way to do this. You should be able to replicate the printing to stderror by just executing the `spotify_web.py`.

aadibajpai replied:
Right, makes sense. I haven't done this myself so I'm not sure but you could google or stackoverflow once I guess, or even ask around in the SO chat :)

JamesVStone replied:
good suggestion. I found a solution where I can override the log function in the `HTTPServer` class. In accordance with this I have removed the reroute of stderror from `spotify.py`.

aadibajpai commented on SwSpotify/spotify.py:
```diff
@@ -112,13 +114,31 @@ def get_info_mac():
+            return get_info_mac()
+        else:
+            return get_info_linux()
+    except SpotifyClosed:
```
You probably want to catch SpotifyNotRunning since if the desktop app is open but the music is playing on the Web Player then SpotifyPaused would be raised, not Closed

aadibajpai commented on SwSpotify/spotify_web.py:
```diff
@@ -0,0 +1,75 @@
+
+
+def server():
+    server_address = ('localhost', 5042)
```
5042 is the port that `swaglyrics -t` runs on

JamesVStone replied:
`5042` is the port number that the `SwagLyrics for spotify` extension was sending the data on. Should I change both the sending and receiving port in this fork and my fork of the extension? @aadibajpai

aadibajpai replied:
Yes, it was 5042 because earlier SwSpotify wasn't an independent module by itself

JamesVStone replied:
changed to port `5043`

JamesVStone commented:
# Expanded documentation

My approach involves attempting to connect to native spotify using library methods and if that fails start a http server. The HTTP server is very lightweight as it is derived from the standard library http.server so their are few additional overheads thus not packages to install. The server will run for between 2 and 3 seconds to determine if the chrome extension is actively pinging the server with information about the currently playing song in spotify web player. The application is now patched to be fully memory managed using thread `.join` syntax to make sure the script is not consuming undue memory. The server does not run forever it only takes two requests (or times out and returns None) one of which would be the HTTP `OPTIONS` request my the probe of the library which is used to send the request and then handling for the `POST` operation where the data is extracted and attached to class state (so to speak). 

I will comment code according to this explanation. 

Thank you for being part of GCI and being such helpful mentors.

JamesVStone commented:
can someone give me some direction on getting the appveyor build to pass :man_facepalming:

aadibajpai commented:
> can someone give me some direction on getting the appveyor build to pass 🤦‍♂

I'll take a look

aadibajpai commented:
> # Expanded documentation
> My approach involves attempting to connect to native spotify using library methods and if that fails start a http server. The HTTP server is very lightweight as it is derived from the standard library http.server so their are few additional overheads thus not packages to install. The server will run for between 2 and 3 seconds to determine if the chrome extension is actively pinging the server with information about the currently playing song in spotify web player. The application is now patched to be fully memory managed using thread `.join` syntax to make sure the script is not consuming undue memory. The server does not run forever it only takes two requests (or times out and returns None) one of which would be the HTTP `OPTIONS` request my the probe of the library which is used to send the request and then handling for the `POST` operation where the data is extracted and attached to class state (so to speak).
> 
> I will comment code according to this explanation.
> 
> Thank you for being part of GCI and being such helpful mentors.

I see, and the extension is constantly sending requests?

aadibajpai commented on tests/test_spotify.py:
```diff
@@ -185,5 +185,44 @@ def test_that_song_function_returns_None_when_error(self):
+        x = song
+        self.assertRaises(SpotifyNotRunning, x)
+
+
```
Sorry I'm just noticing this but you should have tests that actually test song artist info coming up, as in `self.assertEqual(get_info_web(), (song, artist))`

aadibajpai replied:
Currently it doesn't look if something worth testing is being tested

JamesVStone replied:
Sure, thanks for this advice as I was not sure of the required scope for the tests. First time doing unit tests :smile:

aadibajpai replied:
No problem, feel free to ask if you need any help. Unit testing is a useful skill to have :)

aadibajpai commented:
> can someone give me some direction on getting the appveyor build to pass 🤦‍♂

Not 100% sure but it looks like since there's a try except block now, when that SpotifyPaused is caught in the test, the get_info_web function is executed which by itself raises SpotifyPaused which of course isn't caught.

JamesVStone commented:
> > # Expanded documentation
> > My approach involves attempting to connect to native spotify using library methods and if that fails start a http server. The HTTP server is very lightweight as it is derived from the standard library http.server so their are few additional overheads thus not packages to install. The server will run for between 2 and 3 seconds to determine if the chrome extension is actively pinging the server with information about the currently playing song in spotify web player. The application is now patched to be fully memory managed using thread `.join` syntax to make sure the script is not consuming undue memory. The server does not run forever it only takes two requests (or times out and returns None) one of which would be the HTTP `OPTIONS` request my the probe of the library which is used to send the request and then handling for the `POST` operation where the data is extracted and attached to class state (so to speak).
> > I will comment code according to this explanation.
> > Thank you for being part of GCI and being such helpful mentors.
> 
> I see, and the extension is constantly sending requests?

Yes, it sends requests approx once per second.

aadibajpai commented on tests/test_spotify.py:
```diff
@@ -185,5 +185,51 @@ def test_that_song_function_returns_None_when_error(self):
+        """
+        test that get_info_web returns song, artist (or None, None) if Spotify not running
+        """
+        self.assertEqual(get_info_web(), (song, artist))
```
When I said (song, artist) I meant for some song artist pair you should patch the function internally and test that it works, like here (https://github.com/SwagLyrics/SwSpotify/blob/8e53ac1edc53b8fe49e4e4b047c1cddadd181a5f/tests/test_spotify.py#L69)

Currently, since you're patching the function you want to test, it will return true to any output that you give to it

JamesVStone replied:
right, that makes more seance

JamesVStone replied:
Is this the kind of test that you are looking for? @aadibajpai.

```python
    @patch('spotify_web.run', return_value={"title": "Darkside", "artist": "Alan Walker"})
    def test_get_info_web(self, mock_spotify_web):
        """
        test that get_info_web works
        """
        x = get_info_web()
        self.assertEqual(x, ("Darkside", "Alan Walker"))
```

aadibajpai replied:
Yep you're on the right track.

aadibajpai commented on tests/test_spotify.py:
```diff
@@ -185,5 +185,51 @@ def test_that_song_function_returns_None_when_error(self):
+        test that test artist function returns None when the get_info_web function will return an error
+        """
+        x = artist
+        self.assertRaises(SpotifyNotRunning, x)
```
I don't think this test is relevant to the Web part since when the artist function is called, the exception would be raised by the native function and not the web function

aadibajpai commented on tests/test_spotify.py:
```diff
@@ -185,5 +185,51 @@ def test_that_song_function_returns_None_when_error(self):
+        x = artist
+        self.assertRaises(SpotifyNotRunning, x)
+
+    def test_that_song_function_returns_None_when_error(self):
```
Same as above

aadibajpai commented:
You should aim to fully cover testing spotify_web.py especially since it's being added newly and we want it to be future proof. A good way to write tests is to imagine the flow of control and structure tests accordingly. The existing testsuite for the other functions should give you an insight 🙂

aadibajpai commented:
@JamesVStone let me know when it's ready for review

JamesVStone commented:
> You should aim to fully cover testing spotify_web.py especially since it's being added newly and we want it to be future proof. A good way to write tests is to imagine the flow of control and structure tests accordingly. The existing testsuite for the other functions should give you an insight

JamesVStone commented:
acc clicked comment and close :/

aadibajpai commented:
> acc clicked comment and close :/

no worries 😄

JamesVStone commented:
Update
--------------------
The program is now completely synchronous and does not need threading to work. As a result, the program now works better as I have seen in concurrency diagrams. In addition, I have refactored some of the identifier names and improved the interface between the files. I have removed some tests and added some more meaningful and ones that actually prove that the program works (as opposed to just adding coverage).

This is how the programs works now:
+ attempts native connection
+ when this fails it calls  `get_info_web`
+ get info web spawns web server synchronously
+ if the server times out then it will return None and `get_info_web' will raise spotifyClosed
+ On the other hand, If data has been received it will output the respective values
